modules/core/src/test/java/org/apache/ignite/internal/testframework/WorkDirectoryExtension.java

```diff
@@ -0,0 +1,155 @@
+ * <p>
+ * This extension supports both field and parameter injection of {@link Path} parameters annotated with the
+ * {@link WorkDirectory} annotation.
+ * <p>
+ * A new temporary folder is created for every test method and will be located relative to the module,
+ * where the tests are being run, by the following path: "target/work/{@literal <name-of-the-test-method>}".
+ * It is removed after a test has finished running, but this behaviour can be controlled by setting the
+ * {@link WorkDirectoryExtension#KEEP_WORK_DIR_PROPERTY} property to {@code true}, in which case the created folder can
+ * be kept intact for debugging purposes.
+ */
+public class WorkDirectoryExtension implements BeforeEachCallback, AfterEachCallback, ParameterResolver {
+    /**
+     * System property that, when set to {@code true}, will make the extension preserve the created directories.
+     * Default is {@code false}.
+     */
+    public static final String KEEP_WORK_DIR_PROPERTY = "KEEP_WORK_DIR";
+
+    /** Base path for all temporary folders in a module. */
+    private static final Path BASE_PATH = Path.of("target", "work");
+
+    /** {@inheritDoc} */
+    @Override public void beforeEach(ExtensionContext context) throws Exception {
+        Object testInstance = context.getRequiredTestInstance();
+
+        Field workDirField = getWorkDirField(testInstance.getClass());
+
+        if (workDirField == null)
+            return;
+
+        workDirField.setAccessible(true);
+
+        workDirField.set(testInstance, createWorkDir(context));
+    }
+
+    /** {@inheritDoc} */
+    @Override public void afterEach(ExtensionContext context) throws Exception {
+        if (shouldRemoveDir())
+            IgniteUtils.deleteIfExists(BASE_PATH);
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean supportsParameter(
+        ParameterContext parameterContext, ExtensionContext extensionContext
+    ) throws ParameterResolutionException {
+        return getParameterType(parameterContext).equals(Path.class)
+            && parameterContext.isAnnotated(WorkDirectory.class);
+    }
+
+    /** {@inheritDoc} */
+    @Override public Object resolveParameter(
+        ParameterContext parameterContext, ExtensionContext extensionContext
+    ) throws ParameterResolutionException {
+        try {
+            return createWorkDir(extensionContext);
+        }
+        catch (IOException e) {
+            throw new ParameterResolutionException("Error when creating the work directory", e);
+        }
+    }
+
+    /**
+     * Shortcut for extracting the method parameter type from a {@link ParameterContext}.
+     */
+    private static Class<?> getParameterType(ParameterContext parameterContext) {
+        return parameterContext.getParameter().getType();
+    }
+
+    /**
+     * Creates the temporary folder for the given test method.
+     */
+    private static Path createWorkDir(ExtensionContext extensionContext) throws IOException {
+        if (shouldRemoveDir())
```

Comment:
This condition is questionable, we should discuss it

Reply:
On the other hand, this design is good enough for the majority of cases, so nevermind